### PR TITLE
Simulation studio widget

### PR DESCRIPTION
**Pull request by robinr**
Implementation of the Simulation Studio and Simulation Plate Well.

### CODE REVIEW

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    };
+    wrapper = mount(StimulationStudioWidget, {
+      propsData,
+      store,
+      localVue,
+    });
+    const well = wrapper.findAll("circle");
+    expect(well.at(0).attributes("fill")).toStrictEqual("#19AC8A");
+  });
+  test("Given the StimulationStudioWidget is in the range of 0 - 23, When mounted successfully, Then verify that Stimulation Plate Well is teal color and Alphabets A-X is visible", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:52]: I think you chose correctly to just test the first and last wells---I don't think it's necessary to make assertions about every single one of the 24 letters. But if your test only makes assertions about the first and last elements, then that is what your test title should say. That way it is clear what is being tested and what was decided not necessary to explicitly assert (since the inner elements are generated from an array): something like `Then the plate well index 0 is teal with the letter A and plate well index 23 (the last one) is teal with the letter X`

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    const well = wrapper.findAll("circle");
+    const protocol_name = wrapper.findAll(
+      ".span__simulationstudio-plate-well-protocol-location"
+    );
+    expect(well.at(0).attributes("fill")).toStrictEqual("#19AC8A");
+    expect(protocol_name.at(0).text()).toStrictEqual("A");
+    expect(well.at(23).attributes("fill")).toStrictEqual("#19AC8A");
+    expect(protocol_name.at(23).text()).toStrictEqual("X");
+  });
+  test("Given the StimulationStudioWidget is in the range of 24 - 48, When mounted successfully, Then verify that Stimulation Plate Well is teal color(at 0) and blue color(at 23) and Alphabets Y-V is visible", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:72]: As you are writing your tests before writing the next piece of code, one way to help make sure that you are following the TDD guidance of only writing the simplest piece of code to make the current failing test pass is to test the edge cases. For example here, you expect a color change to happen between index 25 and 26 (and also the alphabet to restart), so those would be relevant places to make assertions

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    const well = wrapper.findAll("circle");
+    const protocol_name = wrapper.findAll(
+      ".span__simulationstudio-plate-well-protocol-location"
+    );
+    expect(well.at(0).attributes("fill")).toStrictEqual("#19AC8A");
+    expect(protocol_name.at(0).text()).toStrictEqual("Y");
+    expect(well.at(23).attributes("fill")).toStrictEqual("#005470");
+    expect(protocol_name.at(23).text()).toStrictEqual("V");
+  });
+  test("Given the StimulationStudioWidget is in the range of 48 - 71, When mounted successfully, Then verify that Stimulation Plate Well is blue color(at 0) and yellow color(at 23) and Alphabets W-V is visible", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:93]: similar comment above about testing the border/edge cases of your code logic

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    const well = wrapper.findAll("circle");
+    const protocol_name = wrapper.findAll(
+      ".span__simulationstudio-plate-well-protocol-location"
+    );
+    expect(well.at(0).attributes("fill")).toStrictEqual("#005470");
+    expect(protocol_name.at(0).text()).toStrictEqual("W");
+    expect(well.at(23).attributes("fill")).toStrictEqual("#f9d78c");
+    expect(protocol_name.at(23).text()).toStrictEqual("T");
+  });
+  test("Given the StimulationStudioWidget is in the range of 72 - 96, When mounted successfully, Then verify that Stimulation Plate Well is yellow color(at 0) and orange color(at 23) and Alphabets W-V is visible", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:114]: similar comment above about testing the border/edge cases of your code logic

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      ".span__simulationstudio-plate-well-protocol-location"
+    );
+    expect(well.at(0).attributes("fill")).toStrictEqual("#f9d78c");
+    expect(protocol_name.at(0).text()).toStrictEqual("U");
+    expect(well.at(6).attributes("fill")).toStrictEqual("#df6147");
+    expect(protocol_name.at(6).text()).toStrictEqual("A");
+    expect(well.at(23).attributes("fill")).toStrictEqual("#19AC8A");
+    expect(protocol_name.at(23).text()).toStrictEqual("A");
+  });
+  test("Given the StimulationStudioWidget is null, When mounted successfully, Then verify that Stimulation Plate Well is grey color(at 0)  and No Alphabets are visible", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:139]: similar comment as above about making sure the test title matches what you are testing.  You are only testing the first and last positions, not every position. So instead of "no alphabets are visible", make sure to clarify something like "no letters are visible in the first or last wells"

File: components/playback/controls/StimulationStudioPlateWell.vue [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    circle_y: { type: Number, default: 0 },
+    radius: { type: Number, default: 0 },
+    strk: { type: String, default: "" },
+    protocol_fill: { type: String, default: "" },
+    stroke_wdth: { type: Number, default: 0 },
+    index: { type: Number, default: 0 },
+    protocol_type: { type: String, default: "" },
+  },
+  computed: {
+    /*   0 4  8 12 16 20     In order to speed the rendering its better to pre-compute
```
Comment (eli88fine) [inline components/playback/controls/StimulationStudioPlateWell.vue:57]: this is 130 lines of code instead of ~3 lines that would do the calculations through equations (similar to `PlateNavigator`).  can you elaborate on the benefit you expect in performance that would justify the large increase in the number of lines of code to read through?

**Review by eli88fine — COMMENTED** [review]
Good work overall. Just a few questions and minor corrections that we can go over together in our meeting

File: components/playback/controls/StimulationStudioPlateWell.vue [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+</template>
+<script>
+import PlateWell from "@/components/playback/controls/PlateWell.vue";
+
+export default {
+  name: "StimulationStudioPlateWell",
+  components: {
+    PlateWell,
+  },
+  props: {
```
[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/StimulationStudioPlateWell.vue:43]: `stroke` is a css property type for SVG, a direct utilization matching the name `stroke` in the Vuejs component PlateWell results in css property of the component to be modified, but our business logic is to pass the value of the `stroke` to be applied on an svg element <circle>, so I have just retained the same reference as `strk`, the same applies in case of the attribute of `stroke-width` for the `stroke_wdth` on referring the same name it results in replacement at higher element.

We cannot match the names exactly the same as it results in loosing the flexible re-use of the widget.

File: components/playback/controls/StimulationStudioWidget.vue [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
+
+library.add(faPlusCircle);
+
+export default {
+  name: "StimulationStudioWidget",
+  components: { FontAwesomeIcon, StimulationStudioPlateWell },
+  props: {
+    protocol_code: {
+      type: Array /* This Eslint rule forces to have an function vue/require-valid-default-prop  and creates a function which cannot be code covered */,
```
[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/StimulationStudioWidget.vue:62]: As discussed in the meeting yesterday I have included additional JEST testcase in which the `propsData` is empty and the coverage on the `default()` function is included 
  `Given the StimulationStudioWidget is having no the prop info for protocol_code, When mounted successfully, Then verify that Stimulation Plate Well is grey color and no Alphabets applied visible`

Additionally I have removed the comment of the of the `Eslint`.

File: pages/playback/controls/stimulation-studio/basic.vue [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    this.protocol_list2 = [];
+    for (let i = 24; i < 52; i++) this.protocol_list2.push(i);
+    this.protocol_list3 = [];
+    for (let j = 52; j < 77; j++) this.protocol_list3.push(j);
+    this.protocol_list4 = [];
+    for (let k = 78; k < 95; k++) this.protocol_list4.push(k);
+    for (let l = 0; l < 9; l++) this.protocol_list4.push(l);
+  },
+  methods: {
+    tealcolor() {
```
[the remark this answers is not in the document]
Comment (robinr) [reply pages/playback/controls/stimulation-studio/basic.vue:55]: Corrected.

File: components/playback/controls/StimulationStudioWidget.vue [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+import { faPlusCircle } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
+
+library.add(faPlusCircle);
+
+export default {
+  name: "StimulationStudioWidget",
+  components: { FontAwesomeIcon, StimulationStudioPlateWell },
+  props: {
+    protocol_code: {
```
[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/StimulationStudioWidget.vue:61]: Modified to `protocol_codes`

File: pages/playback/controls/stimulation-well/x-y-offset.vue [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<template>
+  <div style="top: 111px; left: 88px; position: absolute">
+    <ComponentToTest
+      :classname="'plate_0'"
+      :protocol_type="''"
```
[the remark this answers is not in the document]
Comment (robinr) [reply pages/playback/controls/stimulation-well/x-y-offset.vue:5]: Included the protocol alphabet `A` and also modified the color to `#df6147`. The VRT is now modified.

File: tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    const protocol_name = wrapper.find(
+      ".span__simulationstudio-plate-well-protocol-location"
+    );
+    const well_circle = wrapper.find("circle");
+
+    expect(protocol_name.text()).toStrictEqual("'A'");
+    expect(well_circle.attributes("fill")).toStrictEqual("'#19AC8A'");
+    expect(well_circle.attributes("stroke")).toStrictEqual("'#FFFFFF'");
+  });
+  test("Given that a protocol type is A is set index -1  and stroke_wdth = 4 as propsData, When the mounted successfully, Then validate that protocol Text A and color as provided is applied on the circle with white cirlce of 4px", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js:76]: Following computed API's
a) `computed_top`
b) `computed_left`
c) `computed_protocol_top`
d) `computed_protocol_left`

All contain a simple `switch-case` structure which suffices all the required business logic condition, but when we run the eslint it fires up an error called as the follows:-

Expected to return a value in "computed_top" computed property            vue/return-in-computed-property

To satisfy this rule without it forces to introduces the value of return outside `switch-case` which is `return 0`.

NOW after introducing the `return 0` theres is another rule which was introduced in CI build to make 100% code coverage.

This makes us to introduce a testcase in JEST just to cover the lines of code which was introduced due to `Eslint` rule.

So the JEST testcases with `index: -1` is needed.

Currently, there are so many many rules into the Frontend development, which just forces to add un-necessary code due to eslint and on the other hand the code coverage makes the code coverage and TDD is on overdrive, which really is un-necessary.

First the Developer is informed before the ESLINT rules are applied and with a reason and root cause definition, but in our project its almost never practiced.

File: tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+describe("StimulationStudioPlateWell.vue", () => {
+  beforeAll(async () => {
+    // note the store will mutate across tests, so make sure to re-create it in beforeEach
+    const storePath = `${process.env.buildDir}/store.js`;
+    NuxtStore = await import(storePath);
+  });
+
+  beforeEach(async () => {
+    store = await NuxtStore.createStore();
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js:22]: Removed

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
+import { mount } from "@vue/test-utils";
+import StimulationStudioWidget from "@/components/playback/controls/StimulationStudioWidget.vue";
+import { StimulationStudioWidget as DistComponentToTest } from "@/dist/mantarray.common";
+import Vuex from "vuex";
+import { createLocalVue } from "@vue/test-utils";
+
+let wrapper = null;
+
+const localVue = createLocalVue();
+localVue.use(Vuex);
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:10]: Removed

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    const storePath = `${process.env.buildDir}/store.js`;
+    NuxtStore = await import(storePath);
+  });
+
+  beforeEach(async () => {
+    store = await NuxtStore.createStore();
+  });
+
+  afterEach(() => wrapper.destroy());
+  test("When mounting StimulationStudioWidget from the build dist file, Then it loads successfully", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:26]: fixed

File: tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  });
+
+  afterEach(() => wrapper.destroy());
+  test("When mounting StimulationStudioPlateWell from the build dist file, Then it loads successfully", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioPlateWell.spec.js:26]: fixed

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    };
+    wrapper = mount(DistComponentToTest, {
+      propsData,
+      store,
+      localVue,
+    });
+    const well = wrapper.findAll("circle");
+    expect(well).toHaveLength(24);
+  });
+  test("Given the StimulationStudioWidget is in the range of 0 - 25, When mounted successfully, Then verify that Stimulation Plate Well is teal color and Alphabets A-X is visible", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:38]: removed the section about the letters, as there was another one validating the similar below, I removed the assertion, but forgot to modify the title.

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 1 above)
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:52]: No, such detailed `spoon feeding title cannot be written`, its a basic expectation of a developer. 

There is saying:-
`Code is like humor. When you have to explain it, it’s bad.` – Cory House

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 2 above)
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:72]: Not essentially required VRT covers a good amount of all scenarios.

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 3 above)
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:93]: Not essentially required VRT covers a good amount of all scenarios.

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 4 above)
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:114]: Not essentially required VRT covers a good amount of all scenarios.

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 5 above)
Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:139]: modified

File: components/playback/controls/StimulationStudioPlateWell.vue (hunk 6 above)
Comment (robinr) [reply components/playback/controls/StimulationStudioPlateWell.vue:57]: Done.

**Review by robinr — COMMENTED** [review]
Completed most of the PR review comments.

File: components/playback/controls/StimulationStudioPlateWell.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply components/playback/controls/StimulationStudioPlateWell.vue:43]: could you say `stroke_color` instead of `strk` then? it can be very confusing for someone to try to figure out what a variable name is if all the vowels have been removed from the word

how would `stroke_width` (with an underscore for the prop name) conflict with `stroke-width` (with a hyphen, which is the CSS property)?

[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/StimulationStudioPlateWell.vue:43]: `PlateWell` is an atomic widget, modifying on that component will create a large amount of files which are built on top of that.

Why didn't you find it different when PlateWell was reviewed ? 

This cannot be done it will create lots of changes.

[the remark this answers is not in the document]
Comment (eli88fine) [reply components/playback/controls/StimulationStudioPlateWell.vue:43]: I'm not asking you to modify `PlateWell` right now, just `StimulationPlateWell` to stop the issue from cascading further in the library.

Human reviewers often miss things the first time reviewing something.  That is why we use automated linting and other tools for as much as possible to keep our code clean, readable and following best practices. It is well documented that human reviewers can also miss bugs, which is why proper automated testing and Test Driven Development is another important aspect of maintaining a high quality codebase.

[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/StimulationStudioPlateWell.vue:43]: ok... will work on the same.

File: tests/unit/components/playback/controls/StimulationStudioWidget.spec.js (hunk 1 above)
Comment (eli88fine) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:52]: Yes, if someone wants to understand the implementation of a test, then the code should speak for itself when the developer is reading the code, and not require large amounts of inline comments to understand the implementation.

But the goal of descriptive names for Classes, Methods, Functions, and Test Cases is that someone can look at that name and immediately understand what the goal of that section of code is, without spending time looking through the implementation (unless the task they are working on does require a deeper understanding of the implementation).  

Common estimates are that developers spend more than 10X as much time reading code than they spend writing it. So creating clear titles that allow someone reading the code to skip reading the implementation of a specific function (in many cases...of course sometimes the implementation details itself are what the person is trying to understand) is very important to save overall development time.

When a developer is working, they may accidentally impact existing functionality of the codebase.  By following this principle `When a unit test fails, I want to understand what I broke—without reading the test code.`, it will save the developers countless hours because they can usually quickly see from the test title what they broke and then see how the unit of code they just added caused the functionality to break, and then go about correcting it.  If they need to stop, open up the spec file containing the test that failed, and read through the implementation details of of a test to find out what it was supposed to be testing, then that adds dramatically to the development time.

The importance of the principle of clearly naming test cases is why there are numerous online articles describing different potential ways to name test cases.  If you think that there is a better way to name our test cases, please let me know and we can discuss.  But clear test titles are extremely important, and inaccurate test titles (like this one) are dangerous because if this test fails as someone is working, then they would initially assume that the test is doing something different than it actually is, which would slow down their attempts to resolve the issue.

https://qualitycoding.org/unit-test-naming/
https://www.petrikainulainen.net/programming/testing/writing-clean-tests-naming-matters/
https://medium.com/practical-software-testing/write-better-test-cases-titles-e2989d614b5b

Comment (robinr) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:52]: Sorry, I don't agree how come a developer tries to learn everything in the world via only testcase is nemesis.  What happens to the requirements ? is reading requirements the duty of the developer or not ... if you are not willing to accept let it be.

Comment (eli88fine) [reply tests/unit/components/playback/controls/StimulationStudioWidget.spec.js:52]: Requirements documents are used for the Story currently being worked on.  If your Story ends up interacting with code written 5 months or 2 years ago, you are not going to look up the requirements from 2 years ago for a component to understand how it works.  You need to look at something in the codebase.

One of the primary purposes of test case is to serve as a representation of the requirements that stays in the codebase and is easy for anyone to view at any time if they want to understand the purpose and functionality of a component that was created earlier by someone else.

https://capgemini.github.io/development/unit-tests-as-documentation/

https://www.softwaretestinghelp.com/unit-testing/
Documentation availability: Developers who are looking into the functionality at a later stage can refer to the unit testing documentation and can easily find the unit test interface and correct or work fast and easily.

**Review by eli88fine — APPROVED** [review]
